doc/reference/storage_drivers.md

```diff
@@ -55,12 +55,14 @@ Therefore, the first instance takes longer to create than subsequent ones.
 ### Optimized instance transfer
 
 Btrfs, ZFS and Ceph RBD have an internal send/receive mechanism that allows for optimized volume transfer.
-LXD uses this mechanism to transfer instances and snapshots between servers.
+This mechanism requires snapshots to be present and included in the transfer.
+It then transfers only the differences between the snapshots and the main volume, instead of sending the full volume.
+If no snapshots are present, the optimized transfer does not have an advantage and is equivalent to sending the full volume.
```

Comment:
Do the snapshots need to be part of the transfer? (But will we then really save much if a user creates a snapshot right before the transfer?)
Or do they just need to exist on the source server? (But they would need to be copied to the target server as well or the diff won't make sense ...)
Or do we just need one snapshot on the target server? (But how does the optimized transfer work for the first copy - not refresh - then?)

Reply:
Good questions.

The way optimized transfer works is to use the underlying storage driver's native volume send/receive functionality (this is what optimized refers to as its generally quicker than using rsync, even if sending the whole volume).

So for example:

```
lxc storage create zfs1 zfs
lxc storage create zfs2 zfs
lxc launch images:ubuntu/jammy c1 -s zfs1
```

Perform optimized full initial copy to c2 on zfs2 pool:

```
time lxc copy c1 --refresh -s zfs2
real	0m0.793s
```

Then perform a refresh to c2 again and we can see part of the issue described on https://github.com/lxc/lxd/issues/11308
It takes the same amount of time again.

```
time lxc cp c1 c2 --refresh -s zfs2
real	0m0.802s
```

Now lets add a snapshot to the source and refresh again, it takes the same amount of time again because the snapshot is sent to the source and then the (minimal) difference between the snapshot and the main volume is sent:

```
lxc snapshot c1
time lxc cp c1 c2 --refresh -s zfs2
real	0m0.856s
```

Now lets refresh again, now because the snapshot is already on the target, only the differences between the snapshot and the main volume will be sent, so should be quicker:

```
time lxc cp c1 c2 --refresh -s zfs2
real	0m0.263s
```

Equally, taking another snapshot and then refreshing again, sends the new snapshot as only the difference between that and the first snapshot, and then the main volume as the differences between the 2nd snapshot and it.

```
time lxc cp c1 c2 --refresh -s zfs2
real	0m0.271s
```

So optimized transfer relies on periodic snapshots being taken between each refresh, there's no benefit the first time, but after that it helps.

When falling back to rsync transfer, LXD will transparently create a temporary snapshot, but this is just so rsync gets a consistent view of the filesystem during the sync.